File: 56.merge.go
package main

import "sort"

func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(l, r int) bool {
		if intervals[l][0] != intervals[r][0] {
			return intervals[l][0] < intervals[r][0]
		}
		return intervals[l][1] < intervals[r][1]
	})

	result := make([][]int, 0, len(intervals))

	start := intervals[0][0]
	end := intervals[0][1]

	for i := 1; i < len(intervals); i++ {
		currStart := intervals[i][0]
		currEnd := intervals[i][1]

		if currStart <= end {
			end = max(end, currEnd)
			continue
		}

		result = append(result, []int{start, end})

		start = currStart
		end = currEnd
	}

	result = append(result, []int{start, end})

	return result
}

File: 56.merge.c
#include "shdstd.h"

#define MAX(x, y) (((x) > (y)) ? (x) : (y))
int cmp_interval(const void *left, const void *right)
{
	int start_left = (*(int **)left)[0];
	int start_right = (*(int **)right)[0];

	int end_left = (*(int **)left)[1];
	int end_right = (*(int **)right)[1];

	if (start_left != start_right) {
		return start_left - start_right;
	}

	return end_left - end_right;
}

int **merge(int **intervals, int size, int *col_size, int *ret_size, int **ret_col_size)
{
	qsort(intervals, size, sizeof(int *), cmp_interval);

	int **res = malloc(size * sizeof(int *));
	int *res_flat = malloc(size * 2 * sizeof(int));
	int res_size = 0;

	(*ret_col_size) = malloc(size * sizeof(int));

	int interval_start = intervals[0][0];
	int interval_end = intervals[0][1];
	for (int i = 0; i < size; ++i) {
		int curr_start = intervals[i][0];
		int curr_end = intervals[i][1];

		if (curr_start <= interval_end) {
			interval_end = MAX(interval_end, curr_end);
			continue;
		}

		res[res_size] = res_flat + res_size * 2;
		res[res_size][0] = interval_start;
		res[res_size][1] = interval_end;
		(*ret_col_size)[res_size] = 2;
		++res_size;

		interval_start = curr_start;
		interval_end = curr_end;
	}
	res[res_size] = res_flat + res_size * 2;
	res[res_size][0] = interval_start;
	res[res_size][1] = interval_end;
	(*ret_col_size)[res_size] = 2;
	++res_size;

	*ret_size = res_size;

	return res;
}
